packages/material-ui/src/InputBase/Textarea.js

```diff
@@ -1,246 +1,168 @@
 
-  componentDidUpdate() {
-    this.syncHeightWithShadow();
-  }
+  const { current: isControlled } = React.useRef(value != null);
+  const inputRef = React.useRef();
+  const [state, setState] = React.useState({});
+  const shallowRef = React.useRef();
+  const handleRef = useForkRef(ref, inputRef);
 
-  componentWillUnmount() {
-    this.handleResize.clear();
-  }
+  const syncHeight = () => {
+    const input = inputRef.current;
+    const inputShallow = shallowRef.current;
 
-  handleRefInput = ref => {
-    this.inputRef = ref;
+    const computedStyle = window.getComputedStyle(input);
+    inputShallow.style.width = computedStyle.width;
+    inputShallow.value = input.value || props.placeholder || 'x';
 
-    setRef(this.props.textareaRef, ref);
-  };
-
-  handleRefSinglelineShadow = ref => {
-    this.singlelineShadowRef = ref;
-  };
+    // The height of the inner content
+    const innerHeight = inputShallow.scrollHeight;
+    const boxSizing = computedStyle['box-sizing'];
 
-  handleRefShadow = ref => {
-    this.shadowRef = ref;
-  };
+    // Measure height of a textarea with a single row
+    inputShallow.value = 'x';
+    const singleRowHeight = inputShallow.scrollHeight;
 
-  handleChange = event => {
-    this.value = event.target.value;
+    // The height of the outer content
+    let outerHeight = innerHeight;
 
-    if (!this.isControlled) {
-      // The component is not controlled, we need to update the shallow value.
-      this.shadowRef.value = this.value;
-      this.syncHeightWithShadow();
+    if (rowsMin != null) {
+      outerHeight = Math.max(Number(rowsMin) * singleRowHeight, outerHeight);
     }
-
-    if (this.props.onChange) {
-      this.props.onChange(event);
+    if (rowsMax != null) {
+      outerHeight = Math.min(Number(rowsMax) * singleRowHeight, outerHeight);
     }
-  };
-
-  syncHeightWithShadow() {
-    const props = this.props;
-
-    // Guarding for **broken** shallow rendering method that call componentDidMount
-    // but doesn't handle refs correctly.
-    // To remove once the shallow rendering has been fixed.
-    if (!this.shadowRef) {
-      return;
+    outerHeight = Math.max(outerHeight, singleRowHeight);
+
+    if (boxSizing === 'content-box') {
+      outerHeight -=
+        getStyleValue(computedStyle, 'padding-bottom') +
+        getStyleValue(computedStyle, 'padding-top');
+    } else if (boxSizing === 'border-box') {
+      outerHeight +=
+        getStyleValue(computedStyle, 'border-bottom-width') +
+        getStyleValue(computedStyle, 'border-top-width');
     }
 
-    if (this.isControlled) {
-      // The component is controlled, we need to update the shallow value.
-      this.shadowRef.value = props.value == null ? '' : String(props.value);
+    // Need a large enough different to update the height.
+    // This prevents infinite rendering loop.
+    if (state.innerHeight == null || Math.abs(state.innerHeight - innerHeight) > 1) {
+      setState({
```

Comment:
If we memoize syncHeight we better move this into the `setState` so that we don't invalidate the callback on every state change.

```js
setState(prevState => {
  if (prevState.innerHeight == null || Math.abs(prevState.innerHeight - innerHeight) > 1) {
    return { innerHeight, outerHeight }
  }
  return null;
})
```